### DISCUSSION TITLE
Allow opt-out dev-time SSR returned props' JSON-serializability validation

### DISCUSSION THREAD

**Discussion by ivawzh** [vercel/next.js#23958] [Ideas]
# Feature request

Hi team. Thanks for managing this awesome OSS.

## Is your feature request related to a problem? Please describe.

Currently, there is a check at `next dev` on SSR returned `props`' JSON-serializability. It mandatorily throws exceptions with msg `cannot be serialized as JSON. Please only return JSON serializable data type` when returned `props` contains `undefined`, `Date`, etc.

We have to either 

1. manually handle `undefined -> null` and `Date -> string` casting in `SSR`. 
1. Or use some generic transformations like `JSON.parse(JSON.stringify(data))`. 

The first approach adds in a lot of manual work and questionably coupling between data type and SSR fetching process, especially when the data is coming from a 3rd party say Algolia search, Github endpoints, etc. 

The second approach will break our TypeScript inferred typing flow. Arguably we can downgrade the inferred type to type assertion, but that will introduce type-safety risks on changes.

## Describe the solution you'd like

If I have read it correctly, NextJs is always using `JSON.stringify` to serialize the data payload:

- [server](https://github.com/vercel/next.js/blob/2d42b8e0765050bc61cb2f6efadd1f5825a41f6c/packages/next/next-server/server/next-server.ts#L1200)
- [incremental cache](https://github.com/vercel/next.js/blob/2d42b8e0765050bc61cb2f6efadd1f5825a41f6c/packages/next/next-server/server/incremental-cache.ts#L188:L188)
- [api](https://github.com/vercel/next.js/blob/2d42b8e0765050bc61cb2f6efadd1f5825a41f6c/packages/next/next-server/server/api-utils.ts#L287:L287)

Given that `JSON.stringify` does handle date and undefined to string correctly:

- `JSON.stringify(new Date()) // "2020-08-06T10:22:57.118Z"`
- `JSON.stringify({a:undefined}) // '{}'`

That is, the purpose of this existing JSON-serializability validation seems to only act as a server-client API typing guard. However, the current implementation is more like an unnecessary SSR API restriction. I believe it's also bad for performance as it requires adding unnecessary casting.

As I am using TypeScript, I'd prefer to leave the typing concerns to TS if possible. Such as adding the [enhanced version of `InferGetServerSidePropsType` and `InferGetStaticPropsType` mapped types to transform all `Date` to `string` like this](https://gist.github.com/ivawzh/af3d3bcd9f9f1bfa82f49c523d6c3672#file-univeral-transform-ts-L97-L127). 

I understand that not everyone is using TypeScript with NextJs. So that I am proposing an optional skip of the JSON validation.

## Describe alternatives you've considered

If the above mapped type example is not convincing enough to be added to NextJS's out-of-the-box `InferGetServerSidePropsType` and `InferGetStaticPropsType`. I'd still want NextJS allow the opt-out of JSON validation. So that we can take our own risks and implement the mapped type by ourselves.

## Additional context

Related discussions: 
- https://github.com/vercel/next.js/discussions/11498
- https://github.com/vercel/next.js/discussions/11209
- https://github.com/vercel/next.js/issues/14712
- https://github.com/algolia/react-instantsearch/issues/2960

**Comment by realw5** [comment]
I support the second approach described, I am actively have the problem of Algolia with SSR rendering in NextJS not working to it's full potential. 

Since we cannot serialize the current resultState, and SearchState properly, we ended up with a mismatching state when compared client-side that then results in a client-side query to Algiolia which defets the purpose, or at least half of it. The performance here, and # of requests would be dramatically better.

**Comment by charkour** [comment]
Is there a way to opt out of this behavior and let TS do the checking?

**Comment by chernodub** [comment]
Proposed another solution for this https://github.com/vercel/next.js/discussions/45702
